core/src/hash/hasher.rs

```diff
@@ -0,0 +1,360 @@
+        self.len += bytes_len;
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        &self.buffer[..self.len]
+    }
+}
+
+impl<T> Default for CryptoHasher<T, Vec<u8>>
+where
+    T: markers::Sealed,
+{
+    fn default() -> Self {
+        Self::from(Vec::new())
+    }
+}
+
+impl<T, A> From<A> for CryptoHasher<T, A>
+where
+    T: markers::Sealed,
+    A: Accumulator,
+{
+    fn from(mut accumulator: A) -> Self {
+        <A as Accumulator>::reset(&mut accumulator);
+        Self {
+            accumulator,
+            marker: PhantomData,
+        }
+    }
+}
+
+/// Hash functions that can put their results into an output buffer
+/// described by the `Output` type.
+///
+/// # Note
+///
+/// This is an internal trait used to deduplicate some implementations.
+pub trait FinishInto<Output> {
+    fn finish_into(&self, output: &mut Output);
+}
+
+/// Hash functions that can return their hash results as an output buffer
+/// described by the `Output` type.
+///
+/// # Note
+///
+/// - Do not manually implement this trait since it is automatically implemented
+///   as `Finish<O>` for all `T: FinishInto<O>`.
+/// - This is an internal trait used to deduplicate some implementations.
+pub trait Finish<Output> {
+    fn finish(&self) -> Output;
+}
+
+impl<T, A, O> Finish<O> for CryptoHasher<T, A>
+where
+    Self: FinishInto<O>,
+    T: markers::Sealed,
+    A: Accumulator,
+    O: AsMut<[u8]> + Default,
+{
+    fn finish(&self) -> O {
+        let mut output = <O as Default>::default();
+        <Self as FinishInto<O>>::finish_into(self, &mut output);
+        output
+    }
+}
+
+impl<A> FinishInto<[u8; 32]> for Sha2x256Hasher<A>
+where
+    A: Accumulator,
+{
+    fn finish_into(&self, output: &mut [u8; 32]) {
+        super::sha2_256_raw_into(self.accumulator.as_slice(), output)
+    }
+}
+
+impl<A> FinishInto<[u8; 32]> for Keccakx256Hasher<A>
+where
+    A: Accumulator,
+{
+    fn finish_into(&self, output: &mut [u8; 32]) {
+        super::keccak_256_raw_into(self.accumulator.as_slice(), output)
+    }
+}
+
+impl<A> FinishInto<[u8; 32]> for Blake2x256Hasher<A>
+where
+    A: Accumulator,
+{
+    fn finish_into(&self, output: &mut [u8; 32]) {
+        super::blake2_256_raw_into(self.accumulator.as_slice(), output)
+    }
+}
+
+impl<A> FinishInto<[u8; 16]> for Blake2x128Hasher<A>
+where
+    A: Accumulator,
+{
+    fn finish_into(&self, output: &mut [u8; 16]) {
+        super::blake2_128_raw_into(self.accumulator.as_slice(), output)
+    }
+}
+
+impl<A> FinishInto<[u8; 32]> for TwoxHasher<A>
+where
+    A: Accumulator,
+{
+    fn finish_into(&self, output: &mut [u8; 32]) {
+        super::twox_256_raw_into(self.accumulator.as_slice(), output)
+    }
+}
+
+impl<A> FinishInto<[u8; 16]> for TwoxHasher<A>
+where
+    A: Accumulator,
+{
+    fn finish_into(&self, output: &mut [u8; 16]) {
+        super::twox_128_raw_into(self.accumulator.as_slice(), output)
+    }
+}
+
+impl<A> FinishInto<[u8; 8]> for TwoxHasher<A>
```

Comment:
Oh, I see here now that there are multiple `FinishInto<T>` implementations for one type.